NetKVM/Common/ParaNdis-CX.cpp

```diff
@@ -89,16 +89,15 @@ BOOLEAN CParaNdisCX::SendControlMessage(
         if (0 <= m_VirtQueue.AddBuf(sg, nOut, 1, (PVOID)1, NULL, 0))
         {
             UINT len;
-            void *p;
-            /* Control messages are processed synchronously in QEMU, so upon kick_buf return, the response message
-              has been already inserted in the queue */
+
             m_VirtQueue.Kick();
-            p = m_VirtQueue.GetBuf(&len);
-            if (!p)
+            while (!m_VirtQueue.GetBuf(&len))
             {
-                DPrintf(0, ("%s - ERROR: get_buf failed\n", __FUNCTION__));
+                LARGE_INTEGER interval;
+                interval.QuadPart = -1;
+                KeDelayExecutionThread(KernelMode, FALSE, &interval);
```

Comment:
I'm afraid that we are at IRQL = DISPATCH_LEVEL at this point and KeDelayExecutionThread is illegal to call.

Firstly, we're holding a spinlock (`CLockedContext<CNdisSpinLock>` at the top of the function). And even if we found a way around this and were able to release it while waiting, we still might be running in the DPC.

NdisMSleep should work here. It's already used to implement the 'sleep' VirtioLib callback:
https://github.com/virtio-win/kvm-guest-drivers-windows/blob/master/NetKVM/Common/ParaNdis-VirtIO.cpp#L463

Reply:
First of all, NdisMSleep should be used on IRQL < DISPATCH_LEVEL (https://msdn.microsoft.com/en-us/library/windows/hardware/ff563677(v=vs.85).aspx).  NdisStallExecution call work here (https://msdn.microsoft.com/en-us/library/windows/hardware/ff564568(v=vs.85).aspx).

My second concern is that having a loop like this can hang the driver on pause or PM operation. Those can arrive in parallel and the driver might stop the queue. So I think we should sync with state machine as well.

Reply:
Oops, so the 'sleep' VirtioLib callback should be fixed too.

Interestingly, none of the callers of CParaNbuffersdisCX::SendControlMessage do anything with the return value other than debug printout (likely a bug), which makes me wonder if we can make this fully asynchronous, "fire and forget" style. Unless the driver really needs to wait for the host to process any these control messages. Does it?

Reply:
Apparently CX path has no integration with the state machine while the RX [1] and TX [2] both have. While taking into consideration that with vhost=on the loop won't be executed I do think that reading from the device may affect performance a bit while waiting for a reply as commands on the CX path get executed several times a second especially on cores where CX path share the core with TX and RX.

> Unless the driver really needs to wait for the host to process any these control messages. Does it?
From the quick look I made it seems that the return doesn't seem helpful at all as no one is taking it into account beside a debug print. So this might be an option too but it should be taken into consideration that this might change in the future and thus a proper handling mechanism should be introduced as well.

[1] https://github.com/virtio-win/kvm-guest-drivers-windows/blob/0b8bb1abae7132518b052707c03ab9ebbbba47c6/NetKVM/Common/ndis56common.h#L399

[2] https://github.com/virtio-win/kvm-guest-drivers-windows/blob/0b8bb1abae7132518b052707c03ab9ebbbba47c6/NetKVM/Common/ParaNdis-TX.h#L264

Reply:
Regarding async handling - need to take a close look at current configuration changes.